File: vripper-server/src/main/java/tn/mnlr/vripper/web/wsendpoints/DataController.java
package tn.mnlr.vripper.web.wsendpoints;

import lombok.Getter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.handler.annotation.DestinationVariable;
import org.springframework.messaging.simp.annotation.SubscribeMapping;
import org.springframework.stereotype.Controller;
import tn.mnlr.vripper.jpa.domain.Image;
import tn.mnlr.vripper.jpa.domain.LogEvent;
import tn.mnlr.vripper.jpa.domain.Post;
import tn.mnlr.vripper.jpa.domain.Queued;
import tn.mnlr.vripper.services.DataService;
import tn.mnlr.vripper.services.DownloadSpeedService;
import tn.mnlr.vripper.services.GlobalStateService;
import tn.mnlr.vripper.services.VGAuthService;
import tn.mnlr.vripper.services.domain.DownloadSpeed;
import tn.mnlr.vripper.services.domain.GlobalState;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;

@Controller
public class DataController {

  private final VGAuthService VGAuthService;
  private final GlobalStateService globalStateService;
  private final DownloadSpeedService downloadSpeedService;
  private final DataService dataService;

  @Autowired
  public DataController(
      VGAuthService VGAuthService,
      GlobalStateService globalStateService,
      DownloadSpeedService downloadSpeedService,
      DataService dataService) {
    this.VGAuthService = VGAuthService;
    this.globalStateService = globalStateService;
    this.downloadSpeedService = downloadSpeedService;
    this.dataService = dataService;
  }

  @SubscribeMapping("/user")
  public LoggedUser user() {
    return new LoggedUser(VGAuthService.getLoggedUser());
  }

  @SubscribeMapping("/download-state")
  public GlobalState downloadState() {
    return globalStateService.getCurrentState();
  }

  @SubscribeMapping("/speed")
  public DownloadSpeed speed() {
    return new DownloadSpeed(downloadSpeedService.getCurrentValue());
  }

  @SubscribeMapping("/posts")
  public Collection<Post> posts() {
    List<Post> posts = dataService.findAllPosts();
    posts.sort(Comparator.comparing(Post::getAddedOn));
    return posts;
  }

  @SubscribeMapping("/images/{postId}")
  public List<Image> postsDetails(@DestinationVariable("postId") String postId) {
    return dataService.findImagesByPostId(postId);
  }

  @SubscribeMapping("/queued")
  public Collection<Queued> queued() {
    return dataService.findAllQueued();
  }

  @SubscribeMapping("/events")
  public Collection<LogEvent> events() {
    return dataService.findAllEvents();
  }

  @Getter
  public static class LoggedUser {

    private final String user;

    LoggedUser(String user) {
      this.user = user;
    }
  }
}

File: vripper-web/src/main/kotlin/me/mnlr/vripper/web/wsendpoints/DataController.kt
package me.mnlr.vripper.web.wsendpoints

import org.springframework.messaging.handler.annotation.DestinationVariable
import org.springframework.messaging.simp.annotation.SubscribeMapping
import org.springframework.stereotype.Controller
import me.mnlr.vripper.AppEndpointService
import me.mnlr.vripper.entities.ImageDownloadState
import me.mnlr.vripper.entities.LogEvent
import me.mnlr.vripper.entities.PostDownloadState
import me.mnlr.vripper.entities.Thread
import me.mnlr.vripper.formatSI
import me.mnlr.vripper.model.DownloadSpeed
import me.mnlr.vripper.model.GlobalState
import me.mnlr.vripper.model.LoggedUser
import me.mnlr.vripper.repositories.ImageRepository
import me.mnlr.vripper.repositories.LogEventRepository
import me.mnlr.vripper.repositories.PostDownloadStateRepository
import me.mnlr.vripper.repositories.ThreadRepository
import me.mnlr.vripper.services.DownloadSpeedService
import me.mnlr.vripper.services.GlobalStateService
import me.mnlr.vripper.services.VGAuthService

@Controller
class DataController constructor(
    vgAuthService: VGAuthService,
    globalStateService: GlobalStateService,
    downloadSpeedService: DownloadSpeedService,
    appEndpointService: AppEndpointService,
    postDownloadStateRepository: PostDownloadStateRepository,
    imageRepository: ImageRepository,
    threadRepository: ThreadRepository,
    logEventRepository: LogEventRepository
) {
    private val appEndpointService: AppEndpointService
    private val vgAuthService: VGAuthService
    private val globalStateService: GlobalStateService
    private val downloadSpeedService: DownloadSpeedService
    private val postDownloadStateRepository: PostDownloadStateRepository
    private val imageRepository: ImageRepository
    private val threadRepository: ThreadRepository
    private val logEventRepository: LogEventRepository

    init {
        this.vgAuthService = vgAuthService
        this.globalStateService = globalStateService
        this.downloadSpeedService = downloadSpeedService
        this.appEndpointService = appEndpointService
        this.postDownloadStateRepository = postDownloadStateRepository
        this.imageRepository = imageRepository
        this.threadRepository = threadRepository
        this.logEventRepository = logEventRepository
    }

    @SubscribeMapping("/state")
    fun downloadState(): GlobalState {
        return globalStateService.get()
    }

    @SubscribeMapping("/posts")
    fun posts(): Collection<PostDownloadState> {
        return postDownloadStateRepository.findAll()
    }

    @SubscribeMapping("/images/{postId}")
    fun postsDetails(@DestinationVariable("postId") postId: String): List<ImageDownloadState> {
        return imageRepository.findByPostId(postId)
    }

    @SubscribeMapping("/queued")
    fun queued(): Collection<Thread> {
        return threadRepository.findAll()
    }

    @SubscribeMapping("/events")
    fun events(): Collection<LogEvent> {
        return logEventRepository.findAll()
    }
}